### DISCUSSION TITLE
Auto conversion workflow?

### DISCUSSION THREAD

**Discussion by icweaver** [cjdoris/PythonCall.jl#128] [Q&A]
I really appreciate how PythonCall.jl [avoids automatic conversions](https://cjdoris.github.io/PythonCall.jl/stable/pycall/#Automatic-conversion), but I was wondering if there might also be a way to enable this conditionally?

The case I have been running into is that a colleague sends me some pickle or npz file with a bunch of nested, heterogeneous Python objects which I would then like to feed into my Julia pipeline. I've been sprinkling `pyconvert`s or `Py`objects in multiple places before passing the inputs to various Julia functions to get the expected behavior, but wasn't sure if I was approaching this the right way.

I know it would be a hit in performance in the front-end, but is there a way to accomplish this for all python objects that I load (perhaps with [auto conversion rules](https://github.com/cjdoris/PythonCall.jl/issues/125) that we specify), or maybe some workflow more in line with this package's philosophy that I am overlooking?

As a basic example, if I have the following pickle:

```python
import pickle

data = {"entry1":[1, 2, 3], "entry2":{"entry2_1": [3, 2, 1]}}
with open("./data.pkl", "wb") as f:
    pickle.dump(data, f)
```

is there a way to load it in Julia as the following native object?

```julia
using PythonCall
@py import pickle

data = pywith(pickle.load, open("./data.pkl"))

# <auto conversion flag here?>

# Dict{String, Any} with 2 entries:
#  "entry1" => [1, 2, 3]
#  "entry2" => Dict("entry2_1"=>[3, 2, 1])
```

instead of wrapping with:

```julia
PyDict{String, Any}(data)
# PyDict{String, Any} with 2 entries:
#  "entry1" => Py[1, 2, 3]
#  "entry2" => PyDict{Py, Py}('entry2_1'=>[3, 2, 1])
```

So sort of like the PythonCall.jl inverse of [`py"<>"o`](https://github.com/JuliaPy/PyCall.jl/blob/e3e3008ee1b3e449c6382293f58a1f85384beea6/src/pyeval.jl#L190-L192) for returning raw `PyObjects` in PyCall.jl?

**Comment by cjdoris** [comment]
There's nothing to easily achieve this right now.

You can do `pyconvert(Dict{Any,Any}, x)` (or with a union of types if you like) but this isn't deep, in the sense that any nested objects are still converted in the default way.

You could write your own `pyconvertnative(x)` to convert native Python types to native Julia ones by checking the type of `x` (is it None, bool, int, float, list, etc...) and converting it by hand. Could be a nice addition to PythonCall.

**Reply by icweaver** [reply]
Ok, thanks for getting back to me so quickly and for letting me know that I wasn't missing anything obvious. I will definitely have to give this a shot some time

**Reply by cjdoris** [reply]
Thinking about it, a more composible solution could be to define `struct PyNative end` and add a rule so that `pyconvert(PyNative, x)` works like you want.

It's not quite so simple though because lots of code reasonably assumes that converting to a `T` actually returns a `T`, which this wouldn't.

**Reply by icweaver** [reply]
Ah, thanks for the additional insight!

**Comment by hhaensel** [comment]
I started working on this. Would be nice, if someone with deeper knowledge could comment.

```julia
using PythonCall

const PYTYPES = setdiff(getindex.(getfield.(getfield.(methods(ispy), :sig), :types), 2), [Any])

function pyconvert_native(x)
    ispy(x) && PythonCall.pyisinstance(x, pybuiltins.dict) && return pyconvert_dict(x)

    types = setdiff(getfield.(PythonCall._pyconvert_get_rules(pytype(x)), :type), PYTYPES)
    for T in types 
        ans = PythonCall.pytryconvert(T, x)
        PythonCall.pyconvert_isunconverted(ans) || return PythonCall.pyconvert_result(T, ans)
    end
    PythonCall.pyconvert_unconverted()
end

function pyconvert_dict(::Type{T}, d) where T
    jd = pyconvert(T, d)
    kk = pyconvert_native.(keys(jd))
    vv = [ispy(v) ? PythonCall.pyisinstance(v, pybuiltins.dict) ? pyconvert_dict(T, v) : pyconvert_native(v) : v for v in values(jd)]

    T(zip(kk, vv))
end

pyconvert_dict(d) = pyconvert_dict(Dict, d)
```
which allows us to write the following code snippet

```julia
julia> pl = pylist([1, 2, 3]);
julia> ps = pystr("H");
julia> pd = @py dict({"a":"bb", "c": dict({"e": "f"})});

julia> pyconvert_native(pl)
3-element Vector{Int64}:
 1
 2
 3

julia> pyconvert_native(ps)
"H"

julia> pyconvert_native(pd)
Dict{String, Any} with 2 entries:
  "c" => Dict("e"=>"f")
  "a" => "bb"

julia> pyconvert_dict(d)
Dict{String, Any} with 2 entries:
  "c" => Dict("e"=>"f")
  "a" => "bb"

julia> pyconvert_dict(Dict{String, Any}, d)
Dict{String, Any} with 2 entries:
  "c" => Dict{String, Any}("e"=>"f")
  "a" => "bb"

julia> pyconvert_dict(Dict{Any, Any}, d)
Dict{Any, Any} with 2 entries:
  "c" => Dict{Any, Any}("e"=>"f")
  "a" => "bb"
```

**Comment by hhaensel** [comment]
We can also switch to a similar issue #172 for further discussion.

**Reply by hhaensel** [reply]
One more thought:
Could it be reasonable to cache `_pyconvert_get_rules(pytype::Py)` ?

**Comment by hhaensel** [comment]
I implemented caching, so my current code looks like this:
```julia
using PythonCall

const PYTYPES = push!(setdiff(getindex.(getfield.(getfield.(methods(ispy), :sig), :types), 2), [Any]), PyArray)
const PYCONVERT_NATIVE_TYPES_CACHE = Dict{PythonCall.C.PyPtr, Vector{Type}}()

function pyconvert_get_native_ruletypes(x::Py)
    tptr = PythonCall.C.Py_Type(PythonCall.getptr(x))
    get!(PYCONVERT_NATIVE_TYPES_CACHE , tptr) do
        pytype = PythonCall.pynew(PythonCall.incref(tptr))
        ans = setdiff(getfield.(PythonCall._pyconvert_get_rules(pytype), :type), PYTYPES)
        PythonCall.pydel!(pytype)
        ans
    end
end

function pyconvert_native(x)
    ispy(x) || return x

    PythonCall.pyisinstance(x, pybuiltins.dict) && return pyconvert_dict(x)

    types = pyconvert_get_native_ruletypes(x)
    for T in types 
        ans = PythonCall.pytryconvert(T, x)
        PythonCall.pyconvert_isunconverted(ans) || return PythonCall.pyconvert_result(T, ans)
    end
    PythonCall.pyconvert_unconverted()
end

function pyconvert_dict(::Type{T}, d) where T
    jd = pyconvert(T, d)
    kk = pyconvert_native.(keys(jd))
    vv = [ispy(v) ? PythonCall.pyisinstance(v, pybuiltins.dict) ? pyconvert_dict(T, v) : pyconvert_native(v) : v for v in values(jd)]

    T(zip(kk, vv))
end

pyconvert_dict(d) = pyconvert_dict(Dict, d)
```
If this approach looks ok, I would try to cover arrays in a similar manner to dicts.
@cjdoris BTW, is there a reason that `ispy(PyArray) == false`

**Comment by hhaensel** [comment]
With this in place, the following two lines add support for (automatic) DataFrame conversion:
```julia
# add DataFrame conversion rule
using DataFrames

PythonCall.pyconvert_rule_pandasdataframe(::Type{DataFrame}, x::Py) = DataFrame(PyTable(x))
PythonCall.pyconvert_add_rule("pandas.core.frame:DataFrame", DataFrame, PythonCall.pyconvert_rule_pandasdataframe, PythonCall.PYCONVERT_PRIORITY_NORMAL)
```
so
```julia
julia> pd_df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns = pylist("ABC"))
Python DataFrame:
   A  B  C
0  1  2  3
1  4  5  6

julia> pyconvert_native(pd_df)
2×3 DataFrame
 Row │ A      B      C     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      2      3
   2 │     4      5      6
```

**Reply by hhaensel** [reply]
@cjdoris Meanwhile I also covered arrays and implemented support for AbstractDict types during conversion, if you are interested, I can prepare a PR.

**Reply by cjdoris** [reply]
Thanks, I've been thinking a lot lately about how various aspects of PythonCall work, including conversion. I'm considering (a) removing `pyconvert` and just overloading `Base.convert` instead; and (b) converting to native types by default (e.g. `list` to `Vector`, `numpy.ndarray` to `Array`) instead of wrapper types. This would be much more similar to `PyCall`'s behaviour, with the exception that we still won't automatically convert anything.

**Reply by bryaan** [reply]
@cjdoris Any updates on your auto conversion ideas?

**Reply by cjdoris** [reply]
Yeah sorry I've not had loads of time for PythonCall dev lately, and what time I do have has mostly been thinking about v1, trying to stabilise and finalise the API. This will include a more general `pyconvert` which will support to-native conversion but it's a way off. As a result I don't really want to add new functionality right now - but it looks like you can achieve what you want externally anyway, which is good.

The new package extensions stuff is going to be really useful for conversion - it will allow us to add conversion rules for all sorts, such as `DataFrame`s in your example.

**Reply by hhaensel** [reply]
Good to hear, really looking forward to the next version. Feel free to use any of my code above, of course, and let me know if you need help.
Meanwhile I had a python dict including python lists, which needed another definition - just in case someone reads this and want to use my solution before the next version comes out.

```julia
pyconvert_get_native_ruletypes(::PyList{T}) where T = [Vector{T}, Vector]
```

**Comment by hhaensel** [comment]
@cjdoris Did you meanwhile implement a recursive conversion to a native julia type during your big refactoring?

**Reply by cjdoris** [reply]
Not yet but there are changes to `pyconvert` in the works that should allow this - one day you'll be able to do `pyconvert(Any, x, wrap=false)` which will avoid wrapper types.
